### ISSUE TITLE
Please add multi-threaded copy support

### ISSUE THREAD

**Issue by TechyAditya** [files-community/Files#7275]
### What's the Problem?

I was comparing both windows explorer and Files app copy speed of 1600 files. Here are the results. 
![image](https://user-images.githubusercontent.com/72063181/145748124-f171f574-ee91-41d6-bf73-85e27e761290.png)
Files app took 43 seconds to copy 1600 files of 300 kb each
![image](https://user-images.githubusercontent.com/72063181/145748156-0819afe6-0ec7-4efb-906e-3a50a92b374b.png)
Whereas windows just took 5-6 seconds to fully copy the folder. 

### Solution/Idea

Add customizable multi-threaded support in the Files app. I want to set it as 16 or 32 threads (especially when I am copying from google drive to the desktop, multi-threading is essential). 

### Alternatives

You may use robocopy to copy files fast and multi-threaded (worked for my 150 Mbps internet, google drive downloads maxed 145 Mbps, normally it just gives 20 Mbps). 

You may also include file explorer's file handling option in settings to copy files, etc. 

### Priorities

| Capability | Priority |
| Multi-threaded Support | Must |
| Option to choose the number of threads in settings | Should |
| Option to change default file handler to native windows or robocopy | Could |

### Comments

Well, it's an important feature to copy files fast.

**Comment by gave92** [comment]
Files in most cases uses Explorer functions to copy paste. Where did you copy from/to?

**Comment by TechyAditya** [comment]
I was copying from a folder in C drive to another folder in C drive. Just a basic copy-paste benchmarking in an SSD drive.

**Comment by mortenanakin** [comment]
I use Teracopy. Any way of Files let me use Teracopy as the default copy/paste app?

**Comment by t3dium** [comment]
> I use Teracopy. Any way of Files let me use Teracopy as the default copy/paste app?

teracopy isn't multithreaded, it uses a single thread like explorer. use robocopy

**Comment by DarkEden-coding** [comment]
I agree that this is an issue; preferably, multithreading support would exist for all basic file manipulation operations, ie, copy, move, deletion, and transfer from network storage. I'll look into the code later today to see how easy this would be to implement directly into Files.

**Comment by DarkEden-coding** [comment]
Hello, I have made a fork that lets a user choose to use robocopy for copy/move/perma-delete of files. In my testing, it seems significantly faster, ie, 5-10 x, depending on the situation. Should I make a pr or is more discussion required first? Sorry for my ignorance about this. @yaira2 

So far, I have two main functions: one that handles copy/move and another for permanent deletion. The permanent delete one moves files to a temp folder, then uses robocopy to "copy" a blank folder onto that temp folder to remove the files fully. 

I also have settings for the number of threads, switching from Windows shell to robocopy and back, and how large each robocopy batch is (this is to prevent terminal commands from being too long with huge amounts of files). The tradeoff is that the progress updating is more granular the larger the batch number is, but speed is faster the larger the batch is)

**Comment by yair100** [comment]
@DarkEden-coding thank you for reaching out. Is it possible for your implementation to also work with other file operation handlers like TeraCopy?

**Comment by DarkEden-coding** [comment]
@yaira2  It currently uses the shell/terminal api of robocopy, so it would not be a drop-in replacement to use terracopy or others. I'm not sure how file operation handler registration and access works, as I do not use it myself (I think I saw in another issue a different file explorer esc project that does this?). I understand that some users would like to use terracopy, but I'm not sure of the benefits of using terracopy rather than robocopy, as robocopy is multithreaded. Let me know if I'm missing something!

**Comment by Josh65-2201** [comment]
I would assume based on https://github.com/files-community/Files/issues/7518 people want Teracopy because it's what they found to replace Windows File Explorer that still has an interface. If it's fast, reliable and has/can have hash verifying of the files copied then it might be fine to not need Teracopy option

**Comment by TechyAditya** [comment]
Well, teracopy or robocopy, the main priority is to allow crud operations on a large number of files. I use OneDrive mapped as a network drive using rclone. On my wifi bandwidth of 200 Mbps, if I use this app, it goes to 20 Mbps on average if there are lots of files. If I use robocopy, which is Windows' inbuilt api btw, I get 200 Mbps bandwidth maxed out for the same large number of files. I used ChatGPT to generate commands for me to copy the required files. The regular users only care about things running fast and snappy, and do not care underlying algorithm or the tools used. 

Reference: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/robocopy

**Comment by DarkEden-coding** [comment]
Yeah, I agree. I still have to test my Robocopy implementation for cloud drives. I don't have super easy access to large cloud storage, but as long as Robocopy supports cloud drives (it seems to), this should as well. You can find my current implementation at:

https://github.com/DarkEden-coding/Files/tree/Robocopy-Support

You'll have to build it yourself for now, still some ui issues I'm ironing out for now, not sure if settings should be in the advanced tab or the general tab.

**Comment by Josh65-2201** [comment]
If robocopy works just a good as now, I don't see a reason to have a setting. If it's decided to have public testing then it should be under experimental in advanced.

**Comment by DarkEden-coding** [comment]
Could be a compatibility thing for very low power machines but I suppose they can just lower thread count if it shows their system too much.

**Comment by DarkEden-coding** [comment]
For thread count and batch size settings should they be in advanced or general?

**Comment by Josh65-2201** [comment]
This is still in the planning phase so I would wait for Yair to respond before changing anything but if robocopy does one file per CPU thread, having the max be 2/3 or 1/2 of what the system has (make sure that 1 would be minimum) I think would make sense and then a setting isn't needed.

**Comment by TechyAditya** [comment]
I think it's great to have a toggle in the experimental section along with the option to mention the number of threads for now. Once the users liked the implementation and consider it reliable, it can be moved elsewhere in settings, and stay enabled by default (with a safe number of threads equal to the number of CPU cores, or, 2 if it's HDD)

**Comment by DarkEden-coding** [comment]
I'll keep the settings user-controlled for now for testing/beta. What do y'all think about managing Robocopy installation? Should it be user-controlled, or bundled with Files, auto-installed, or something else?

**Comment by Josh65-2201** [comment]
Isn't it build in with Windows?

**Comment by DarkEden-coding** [comment]
Oh, you right, nvm then!

**Comment by DarkEden-coding** [comment]
Here is a benchmark: 1801 images, same drive, each image is 216 kb.

First, without robocopy:

<img width="537" height="347" alt="Image" src="https://github.com/user-attachments/assets/27dcc79a-b145-48d1-b91d-dd648d90097c" />

Then with robocopy (batch size of 40, 6 threads):

<img width="525" height="334" alt="Image" src="https://github.com/user-attachments/assets/aee40dfc-c79a-4290-b443-fc96643cbadc" />

For sure needs more testing, but it seems to be WORLDS faster!

**Comment by Josh65-2201** [comment]
What does batch size change in the copy process?

**Comment by DarkEden-coding** [comment]
It increases the number of files sent each time to the robocopy command (files split into chunks for each robocopy command), I tried to send all files selected at once but I reached the shell command length limit with all the args, there is probably a way to bypass/fix this but the batching was the fast fix for now. It would also be possible to do auto batching based on the file name lengths.

**Comment by Josh65-2201** [comment]
So it isn't wasting your time, I would recommend waiting for yair to approve using robocopy but there are many issues for me, using 16 out of 24 threads and batch size 20.
- 117 out of 1317 items were missing from the paste but status center report it was successful. This happens with any operation, not all selected items are added to the queue.

- Selecting a folder for copying and then pasting into my downloads folder doesn't make the copied folder and just put the files directly in my downloads folder

- Copying an item and pasting in the same directory doesn't create a ` - copy` version of it

- Conflict dialog set to replace on a cut item doesn't delete the file in the original location or replace the file in the paste location.

- Conflict dialog set to generate new name doesn't create the file with the Windows default `file name (Number)`

**Comment by yair100** [comment]
> @yaira2 It currently uses the shell/terminal api of robocopy, so it would not be a drop-in replacement to use terracopy or others. I'm not sure how file operation handler registration and access works, as I do not use it myself (I think I saw in another issue a different file explorer esc project that does this?). I understand that some users would like to use terracopy, but I'm not sure of the benefits of using terracopy rather than robocopy, as robocopy is multithreaded. Let me know if I'm missing something!

@DarkEden-coding thank you for the response. I mentioned TeraCopy only as an example of another operation handler. I should also clarify that there are actually two separate requests:
- Multi-threaded file operations
- Support for third-party operation handlers

**Comment by yair100** [comment]
Regarding replacing our current implementation with RoboCopy, we'll need to do a lot of testing, but I think it's definitely worth exploring.

**Comment by DarkEden-coding** [comment]
> > [@yaira2](https://github.com/yaira2) It currently uses the shell/terminal api of robocopy, so it would not be a drop-in replacement to use terracopy or others. I'm not sure how file operation handler registration and access works, as I do not use it myself (I think I saw in another issue a different file explorer esc project that does this?). I understand that some users would like to use terracopy, but I'm not sure of the benefits of using terracopy rather than robocopy, as robocopy is multithreaded. Let me know if I'm missing something!
> 
> [@DarkEden-coding](https://github.com/DarkEden-coding) thank you for the response. I mentioned TeraCopy only as an example of another operation handler. I should also clarify that there are actually two separate requests:
> 
> * Multi-threaded file operations
> * Support for third-party operation handlers

Got it, I don't believe I have the experience necessary for doing third-party operation handlers.

> So it isn't wasting your time, I would recommend waiting for yair to approve using robocopy but there are many issues for me, using 16 out of 24 threads and batch size 20.
> 
> * 117 out of 1317 items were missing from the paste but status center report it was successful. This happens with any operation, not all selected items are added to the queue.
> * Selecting a folder for copying and then pasting into my downloads folder doesn't make the copied folder and just put the files directly in my downloads folder
> * Copying an item and pasting in the same directory doesn't create a ` - copy` version of it
> * Conflict dialog set to replace on a cut item doesn't delete the file in the original location or replace the file in the paste location.
> * Conflict dialog set to generate new name doesn't create the file with the Windows default `file name (Number)`

@Josh65-2201  I fixed some of your issues in my latest commit. Folder handling was all messed up, so now copy and move operations should create folders correctly. It's possible that my changes also fixed your missing files, but I am not sure, as I could not recreate that issue.

I looked into it, and it seems that Robocopy itself cannot create files with - copy appended to the end of the name. I'm not sure how to address this other than just using the shell interface in these instances. In the same light, renaming for file conflicts seems troublesome in robocopy, so perhaps we can robocopy all other files except these? I'll work on getting the cut conflict dialog working correctly tmr, thank you for testing out my changes!

I also removed the manual batch setting for auto batch size calculation based on file name lengths, so yet faster and clearer for the end user!

**Comment by TechyAditya** [comment]
> - 117 out of 1317 items were missing from the paste but status center report it was successful. This happens with any operation, not all selected items are added to the queue.

I was not having this issue when I was using Robocopy in CLI. Maybe there are some bugs in the implementation. This needs to be fixed or it is pointless to use it. 

Another alternative approach is to use the explorer.exe shell to do file operations for now. This defeats the whole point of using the Files app, that is, to avoid explorer.exe.

**Comment by DarkEden-coding** [comment]
> > * 117 out of 1317 items were missing from the paste but status center report it was successful. This happens with any operation, not all selected items are added to the queue.
> 
> I was not having this issue when I was using Robocopy in CLI. Maybe there are some bugs in the implementation. This needs to be fixed or it is pointless to use it.
> 
> Another alternative approach is to use the explorer.exe shell to do file operations for now. This defeats the whole point of using the Files app, that is, to avoid explorer.exe.

@TechyAditya, I couldn't reproduce the missing files issue myself, it for sure needs more testing, but for me, it copies/moves all files and folders correctly. I also changed how files are preprocessed/grouped for robocopy after Josh's message, so it's possible that fixed the bug that caused their issue.

**Comment by Josh65-2201** [comment]
I will test with the the changes.

<img width="834" height="137" alt="Image" src="https://github.com/user-attachments/assets/e83886b7-5e0c-412e-9c0b-227685ad5ff4" />

nitpicks: the settings toggle should just be on or off to match the rest of the settings as it being on already show it will use robocopy with the name, threads should also be under an expanders like `Files settings > General > Startup items` to not make the setting box tall

**Comment by Josh65-2201** [comment]
- Permanently deleting still doesn't do all selected files 277/1319 were left out

- Trying to copy 1000+ selected items doesn't work, it shows a blank status center successful message and pasting doesn't become available.

- Maybe what is causing the above but copying only files (20 selected) isn't working, Status center reports it as successful but nothing is copied, files in a folder that is copied work correctly.

- There some memory leaking issue as copying, permanently deleting then normal deleting (the 277 left form above issue) stuck CPU usage 90% (set to use 8 threads) and memory at 2.5GB after the operation is done. It did this yesterday testing also.

- Folder that have icons and are copied/moved lose there icon until Files is restarted, this is probably a limitation with how Windows loads icons on folders but finding a way to force refresh like using the none robocopy methods could be looked into.

**Comment by yair100** [comment]
> In the same light, renaming for file conflicts seems troublesome in robocopy, so perhaps we can robocopy all other files except these?

@DarkEden-coding that should be okay.

Regarding the setting, I see the use case for changing the number of threads, but is there any reason someone would want to disable robocopy all together?

**Comment by DarkEden-coding** [comment]
@yaira2 I don't see much of a reason once testing is done. I'm currently using it for testing back and forth for functionality verification.

**Comment by yair100** [comment]
Fair enough, thank you for all your work!

**Comment by Josh65-2201** [comment]
If implemented, It should be a toggle in a preview build for a while so that people can go back to what they are use to, if it doesn't  work correctly

**Comment by DarkEden-coding** [comment]
@Josh65-2201 I can't seem to reproduce your issues with copy/move and perma delete

Here is src folder (2000 random files of random endings and random content)

<img width="1215" height="766" alt="Image" src="https://github.com/user-attachments/assets/969f5f8c-cd0a-4eb3-b8b8-5af4b6e374a6" />

Here is the dest folder after copy (after restart so no success icon):

<img width="1214" height="771" alt="Image" src="https://github.com/user-attachments/assets/cf063a8e-7a79-465a-9c8a-b7f3ec51ae86" />

Here is a basic verification script result (based on file names and sizes)

<img width="811" height="453" alt="Image" src="https://github.com/user-attachments/assets/1f8fef24-e626-440a-ae7d-57e6fba6f178" />

Hash-based verification:

<img width="823" height="251" alt="Image" src="https://github.com/user-attachments/assets/0a752279-ae8d-40f6-b9aa-dd193245e4e8" />

And this is a different folder that I copied the 2k files into, then perma deleted

<img width="1204" height="762" alt="Image" src="https://github.com/user-attachments/assets/7075aa21-468d-41f6-bf05-c09a9678cc40" />

Could you show a screenshot of your file structure? I'm very confused why I'm getting such different results.

**Comment by DarkEden-coding** [comment]
The strange pre-copy behavior is in the base version of Files. I'll investigate if it's exacerbated by my changes, 

Here is a screenshot directly after pressing just the copy button in base Files;

<img width="1219" height="345" alt="Image" src="https://github.com/user-attachments/assets/ef0f6cba-52db-42a0-9eaf-3216c0572cf9" />

**Comment by Josh65-2201** [comment]
I'm using just .mp3 formatted `text - text - text` and two folders into one empty one in user folder downloads. I will do a restart in my next tests to make sure it updating the folder display correctly

**Comment by Josh65-2201** [comment]
On the latest commit, running with 8 threads
- I'm still having the issue of the files not all copying, but isn't happening all the time now. Can't test further due to the below issue

- The an issue where Files won't allow the copy button to work on 600+ items but only showed up recently even after a restart it still isn't working now so I'm not sure if it related to changes here.

- This may be because it not updated to main but it very unstable when deleting, large (1360 items) deletions is a lot slower than 4.0.12 and often freezes until killed by task manager.

**Comment by Jack251970** [comment]
Hopefully, the copy performance could be improved

**Comment by yair100** [comment]
@DarkEden-coding have you made any progress on your proof of concept?

**Comment by DarkEden-coding** [comment]
@yaira2 Ive been a bit busy recently so no new progress. I think I need a few other people to test my changes as I cant seem to recreate the issues Josh has been experiencing, ill take another look through the code today to see if theres anything I can find. Thanks for the follow up!

**Comment by SoftwareExplorer6600** [comment]
Hey, how can I test the changes, do I need to sign up or compile a build or something. if not, could someone provide a link to the artifact so I can download it.

**Comment by DarkEden-coding** [comment]
I just recently started working on this again lol, @Gaming-tech-cyber best way is to clone https://github.com/DarkEden-coding/Files and then build it from that source code, https://files.community/docs/contributing/building-from-source has good instructions

**Comment by DarkEden-coding** [comment]
@yair100 I've updated my fork to be up to date with current files changes, I've also done a good amount of testing, any way of moving this along a bit more?

**Comment by yair100** [comment]
That's great to hear! If the issues @Josh65-2201 reported have been resolved, the next step would be to open a PR for review.
